# Data Collection and Data Cleaning

In [60]:
import pandas as pd
import numpy as np

In [61]:
df = pd.read_csv('sample_data.csv')

# 1. Inspect & Understand Data

🔹 df.head()

Purpose: Display the 5 rows of a DataFrame.


In [62]:
# Example of using the head()
df.head()

Name   Age    Salary         City  Height_cm  Birth_Year Product  \
0   Dave  28.0  150000.0      Chicago      150.0        1992  LAPTOP   
1    Eve  28.0   50000.0     new york        NaN        1992  LAPTOP   
2  Carol  35.0   70000.0      Chicago      190.0        1993  LAPTOP   
3    Eve  28.0   50000.0  los angeles      180.0        1990  LAPTOP   
4    Eve   NaN       NaN  los angeles      150.0        1992  Laptop   

  Purchase_Date  
0           NaN  
1    2022-02-05  
2    2022-03-12  
3    2022-01-01  
4           NaN

🔹 df.tail()

Purpose: Display the last few rows of a DataFrame.


In [63]:
# Example of using the tail()
df.tail()

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
145  Alice   NaN  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   NaN  900000.0     new york      170.0        1993    phone   
147    Eve   NaN  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   NaN   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        NaN        1994   LAPTOP   

    Purchase_Date  
145    2022-03-12  
146           NaN  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01

🔹 df.sample()

Purpose: Return randomly selected rows.

In [64]:
# Example of using the sample()
df.sample()

Name  Age   Salary         City  Height_cm  Birth_Year Product  \
13  Eve  NaN  80000.0  los angeles      180.0        1995   Phone   

   Purchase_Date  
13           NaN

🔹 df.info()

Purpose: Provide a summary of DataFrame structure, data types, and non-null counts.

In [65]:
# Example of using the info()
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Name           150 non-null    str    
 1   Age            94 non-null     float64
 2   Salary         132 non-null    float64
 3   City           150 non-null    str    
 4   Height_cm      128 non-null    float64
 5   Birth_Year     150 non-null    int64  
 6   Product        150 non-null    str    
 7   Purchase_Date  115 non-null    str    
dtypes: float64(3), int64(1), str(4)
memory usage: 9.5 KB


🔹 df.describe()

Purpose: Generate statistical summary of numeric columns.

In [66]:
# Example of using the describe()
df.describe()

Age         Salary   Height_cm   Birth_Year
count  94.000000     132.000000  128.000000   150.000000
mean   29.468085  202575.757576  171.171875  1991.786667
std     4.126239  289272.900582   15.039114     3.075311
min    25.000000   50000.000000  150.000000  1985.000000
25%    25.000000   60000.000000  160.000000  1990.000000
50%    28.000000   75000.000000  170.000000  1993.000000
75%    35.000000  150000.000000  182.500000  1994.000000
max    35.000000  900000.000000  190.000000  1995.000000

🔹 df.shape

Purpose: Return total number of rows and columns.

In [67]:
# Example of using the shape
df.shape

(150, 8)

🔹 df.columns

Purpose: List all column names.

In [68]:
# Example of using the columns
df.columns

Index(['Name', 'Age', 'Salary', 'City', 'Height_cm', 'Birth_Year', 'Product',
       'Purchase_Date'],
      dtype='str')

🔹 df.dtypes

Purpose: Show data type of each column.

In [69]:
# Example of using the dtypes
df.dtypes

Name                 str
Age              float64
Salary           float64
City                 str
Height_cm        float64
Birth_Year         int64
Product              str
Purchase_Date        str
dtype: object

🔹 df.nunique()

Purpose: Count unique values per column.

In [70]:
# Example of using the nunique()
df.nunique()

Name             6
Age              4
Salary           6
City             6
Height_cm        5
Birth_Year       6
Product          6
Purchase_Date    3
dtype: int64

🔹 df.memory_usage()

Purpose: Show memory consumption per column.

In [71]:
# Example of using the memory_usage()
df.memory_usage()

Index             132
Name             1200
Age              1200
Salary           1200
City             1200
Height_cm        1200
Birth_Year       1200
Product          1200
Purchase_Date    1200
dtype: int64

# 2. Handle Missing Values (NaN / None)

🔹 df.isna() / df.isnull()

Purpose: Identify missing values in the DataFrame.

In [72]:
# Example of using the isna()
df.isna()

Name    Age  Salary   City  Height_cm  Birth_Year  Product  \
0    False  False   False  False      False       False    False   
1    False  False   False  False       True       False    False   
2    False  False   False  False      False       False    False   
3    False  False   False  False      False       False    False   
4    False   True    True  False      False       False    False   
..     ...    ...     ...    ...        ...         ...      ...   
145  False   True   False  False      False       False    False   
146  False   True   False  False      False       False    False   
147  False   True   False  False      False       False    False   
148  False   True   False  False      False       False    False   
149  False  False   False  False       True       False    False   

     Purchase_Date  
0             True  
1            False  
2            False  
3            False  
4             True  
..             ...  
145          False  
146           True  
147          False  
148          False  
149          False  

[150 rows x 8 columns]

In [73]:
# Example of using the isnull()
df.isnull()

Name    Age  Salary   City  Height_cm  Birth_Year  Product  \
0    False  False   False  False      False       False    False   
1    False  False   False  False       True       False    False   
2    False  False   False  False      False       False    False   
3    False  False   False  False      False       False    False   
4    False   True    True  False      False       False    False   
..     ...    ...     ...    ...        ...         ...      ...   
145  False   True   False  False      False       False    False   
146  False   True   False  False      False       False    False   
147  False   True   False  False      False       False    False   
148  False   True   False  False      False       False    False   
149  False  False   False  False       True       False    False   

     Purchase_Date  
0             True  
1            False  
2            False  
3            False  
4             True  
..             ...  
145          False  
146           True  
147          False  
148          False  
149          False  

[150 rows x 8 columns]

🔹 df.notna() / df.notnull()

Purpose: Identify non-missing values.

In [74]:
# Example of using the notna()
df.notna()

Name    Age  Salary  City  Height_cm  Birth_Year  Product  Purchase_Date
0    True   True    True  True       True        True     True          False
1    True   True    True  True      False        True     True           True
2    True   True    True  True       True        True     True           True
3    True   True    True  True       True        True     True           True
4    True  False   False  True       True        True     True          False
..    ...    ...     ...   ...        ...         ...      ...            ...
145  True  False    True  True       True        True     True           True
146  True  False    True  True       True        True     True          False
147  True  False    True  True       True        True     True           True
148  True  False    True  True       True        True     True           True
149  True   True    True  True      False        True     True           True

[150 rows x 8 columns]

In [75]:
# Example of using the notnull()
df.notnull()

Name    Age  Salary  City  Height_cm  Birth_Year  Product  Purchase_Date
0    True   True    True  True       True        True     True          False
1    True   True    True  True      False        True     True           True
2    True   True    True  True       True        True     True           True
3    True   True    True  True       True        True     True           True
4    True  False   False  True       True        True     True          False
..    ...    ...     ...   ...        ...         ...      ...            ...
145  True  False    True  True       True        True     True           True
146  True  False    True  True       True        True     True          False
147  True  False    True  True       True        True     True           True
148  True  False    True  True       True        True     True           True
149  True   True    True  True      False        True     True           True

[150 rows x 8 columns]

🔹 df.dropna()

Purpose: Remove rows or columns containing missing values.

In [76]:
df

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
0     Dave  28.0  150000.0      Chicago      150.0        1992   LAPTOP   
1      Eve  28.0   50000.0     new york        NaN        1992   LAPTOP   
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
4      Eve   NaN       NaN  los angeles      150.0        1992   Laptop   
..     ...   ...       ...          ...        ...         ...      ...   
145  Alice   NaN  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   NaN  900000.0     new york      170.0        1993    phone   
147    Eve   NaN  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   NaN   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        NaN        1994   LAPTOP   

    Purchase_Date  
0             NaN  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4             NaN  
..            ...  
145    2022-03-12  
146           NaN  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

In [77]:
# Example of using the dropna()
drop_missing = df.dropna()
display(drop_missing)

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
5      Bob  28.0   60000.0  los angeles      190.0        1992   LAPTOP   
6    Carol  35.0  150000.0      Chicago      150.0        1990    phone   
7    Carol  35.0   60000.0  los angeles      170.0        1990   Tablet   
10    Dave  30.0   70000.0  los angeles      150.0        1992   LAPTOP   
14     Bob  25.0   50000.0      Chicago      150.0        1992  lap top   
24   Frank  35.0   80000.0     New York      170.0        1993    Phone   
25     Eve  30.0   70000.0      Chicago      180.0        1993    phone   
27   Alice  25.0  150000.0      Chicago      180.0        1985    phone   
29   Carol  25.0  150000.0     new york      180.0        1994    Phone   
30   Carol  25.0  150000.0  los angeles      170.0        1990   Laptop   
31     Bob  25.0  150000.0      Chicago      160.0        1985   Tablet   
32    Dave  35.0  150000.0     new york      190.0        1993  lap top   
36   Frank  28.0  150000.0     New York      150.0        1994    phone   
37   Carol  25.0   70000.0      chicago      180.0        1990  lap top   
44     Eve  25.0  900000.0      Chicago      190.0        1990    phone   
45   Alice  35.0   50000.0      chicago      160.0        1995    phone   
50   Frank  25.0   80000.0  Los Angeles      170.0        1994   Tablet   
54     Bob  25.0   80000.0      Chicago      160.0        1993    Phone   
55     Eve  30.0   80000.0      chicago      180.0        1995   Tablet   
60    Dave  35.0  150000.0      chicago      150.0        1993    phone   
67    Dave  30.0  150000.0      Chicago      160.0        1985   LAPTOP   
71   Frank  28.0   80000.0      chicago      190.0        1993    Phone   
72     Bob  25.0   50000.0      Chicago      190.0        1990  lap top   
73    Dave  25.0   80000.0     New York      190.0        1990  lap top   
79     Bob  30.0   70000.0  los angeles      180.0        1994   Tablet   
82    Dave  35.0   60000.0  Los Angeles      190.0        1993   Laptop   
83   Frank  35.0  900000.0     New York      180.0        1993    phone   
84   Frank  25.0  150000.0     new york      180.0        1992  lap top   
86   Alice  28.0   70000.0      Chicago      180.0        1995   Tablet   
88     Eve  30.0   60000.0     New York      190.0        1990   Laptop   
91     Eve  35.0  900000.0  Los Angeles      170.0        1990   Laptop   
92     Bob  25.0   60000.0  los angeles      160.0        1990   Tablet   
93   Alice  25.0  900000.0  los angeles      160.0        1994  lap top   
97     Eve  28.0   60000.0     new york      190.0        1990  lap top   
98   Alice  35.0   50000.0     new york      190.0        1992   Tablet   
106  Carol  35.0  900000.0     new york      150.0        1995    Phone   
109  Carol  25.0   60000.0  los angeles      190.0        1993  lap top   
110  Carol  25.0   50000.0      Chicago      180.0        1994  lap top   
112  Carol  25.0   80000.0  los angeles      150.0        1985   LAPTOP   
113    Eve  35.0   60000.0      chicago      180.0        1985    Phone   
115    Bob  25.0   50000.0      Chicago      150.0        1992    Phone   
116  Alice  25.0   80000.0  Los Angeles      150.0        1990   Laptop   
120    Bob  35.0  150000.0     New York      160.0        1990    Phone   
122  Frank  30.0   80000.0     New York      190.0        1985    phone   
123    Eve  25.0   70000.0     New York      190.0        1985   Tablet   
127  Carol  30.0   60000.0  Los Angeles      190.0        1994    phone   
128  Carol  25.0   60000.0     new york      170.0        1995    phone   
131    Eve  28.0  150000.0      Chicago      180.0        1990   LAPTOP   
132  Frank  28.0   70000.0     New York      170.0        1985    Phone   
133  Carol  35.0   50000.0      Chicago      170.0        1995   Tablet   
135    Eve  25.0  900000.0     

Before we have 150 rows but after use dropna() we only see 144 rows.

🔹 df.fillna()

Purpose: Replace missing values with specified values.

In [78]:
df

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
0     Dave  28.0  150000.0      Chicago      150.0        1992   LAPTOP   
1      Eve  28.0   50000.0     new york        NaN        1992   LAPTOP   
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
4      Eve   NaN       NaN  los angeles      150.0        1992   Laptop   
..     ...   ...       ...          ...        ...         ...      ...   
145  Alice   NaN  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   NaN  900000.0     new york      170.0        1993    phone   
147    Eve   NaN  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   NaN   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        NaN        1994   LAPTOP   

    Purchase_Date  
0             NaN  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4             NaN  
..            ...  
145    2022-03-12  
146           NaN  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

In [79]:
# Example of using the fillna()
fill = df.fillna(0)
display(fill)

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
0     Dave  28.0  150000.0      Chicago      150.0        1992   LAPTOP   
1      Eve  28.0   50000.0     new york        0.0        1992   LAPTOP   
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
4      Eve   0.0       0.0  los angeles      150.0        1992   Laptop   
..     ...   ...       ...          ...        ...         ...      ...   
145  Alice   0.0  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   0.0  900000.0     new york      170.0        1993    phone   
147    Eve   0.0  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   0.0   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        0.0        1994   LAPTOP   

    Purchase_Date  
0               0  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4               0  
..            ...  
145    2022-03-12  
146             0  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

Before in the first row Purchase_Date is NaN but after we use fillna to fill 0 on it, now NaN -> 0.

🔹 df.interpolate()

Purpose: Estimate and fill missing values using mathematical interpolation.

In [80]:
# Example of using the interpolate()
df_interpolate = df.copy()
df_interpolate['Salary'] = df_interpolate['Salary'].interpolate()
print(df_interpolate[['Salary']].head())


     Salary
0  150000.0
1   50000.0
2   70000.0
3   50000.0
4   55000.0


🔹 df.replace(np.nan, value)

Purpose: Replace NaN or other specified values with a new value.

In [88]:
df

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
0     Dave  28.0  150000.0      Chicago      150.0        1992   LAPTOP   
1      Eve  28.0   50000.0     new york        NaN        1992   LAPTOP   
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
4      Eve   NaN       NaN  los angeles      150.0        1992   Laptop   
..     ...   ...       ...          ...        ...         ...      ...   
145  Alice   NaN  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   NaN  900000.0     new york      170.0        1993    phone   
147    Eve   NaN  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   NaN   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        NaN        1994   LAPTOP   

    Purchase_Date  
0             NaN  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4             NaN  
..            ...  
145    2022-03-12  
146           NaN  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

In [ ]:
# Example of using the replace()
replace = df.copy()
replace.replace(np.nan, "Missing", inplace=True)

display(replace)


Name      Age    Salary         City Height_cm  Birth_Year  Product  \
0     Dave     28.0  150000.0      Chicago     150.0        1992   LAPTOP   
1      Eve     28.0   50000.0     new york   Missing        1992   LAPTOP   
2    Carol     35.0   70000.0      Chicago     190.0        1993   LAPTOP   
3      Eve     28.0   50000.0  los angeles     180.0        1990   LAPTOP   
4      Eve  Missing   Missing  los angeles     150.0        1992   Laptop   
..     ...      ...       ...          ...       ...         ...      ...   
145  Alice  Missing  150000.0  Los Angeles     150.0        1994   LAPTOP   
146   Dave  Missing  900000.0     new york     170.0        1993    phone   
147    Eve  Missing  150000.0  Los Angeles     180.0        1990   Laptop   
148   Dave  Missing   70000.0  los angeles     160.0        1993  lap top   
149  Frank     35.0   80000.0     New York   Missing        1994   LAPTOP   

    Purchase_Date  
0         Missing  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4         Missing  
..            ...  
145    2022-03-12  
146       Missing  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

Before in the first row of Purchase_Date is NaN but after use replace it's change to Missing.

🔹 Series.fillna()

Purpose: Replace missing values in a specific column.

In [91]:
df

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
0     Dave  28.0  150000.0      Chicago      150.0        1992   LAPTOP   
1      Eve  28.0   50000.0     new york        NaN        1992   LAPTOP   
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
4      Eve   NaN       NaN  los angeles      150.0        1992   Laptop   
..     ...   ...       ...          ...        ...         ...      ...   
145  Alice   NaN  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   NaN  900000.0     new york      170.0        1993    phone   
147    Eve   NaN  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   NaN   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        NaN        1994   LAPTOP   

    Purchase_Date  
0             NaN  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4             NaN  
..            ...  
145    2022-03-12  
146           NaN  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

In [ ]:
# Example of using series fillna()
ser_fill = df['Age'].fillna(df['Age'].mean())
ser_fill

0      28.000000
1      28.000000
2      35.000000
3      28.000000
4      29.468085
         ...    
145    29.468085
146    29.468085
147    29.468085
148    29.468085
149    35.000000
Name: Age, Length: 150, dtype: float64

In the row 5 befor is NaN but after use series fillna it appear 28.00(average age) instead.

# 3. Remove Duplicates

🔹 df.duplicated()

Purpose: Identify duplicate rows.

In [82]:
# example of using the duplicated()
dup = df.duplicated()
dup

0      False
1      False
2      False
3      False
4      False
       ...  
145    False
146    False
147    False
148    False
149    False
Length: 150, dtype: bool

🔹 df.drop_duplicates()

Purpose: Remove duplicate rows from dataset.

In [83]:
# Example of using the drop_duplicates()
df.drop_duplicates()

Name   Age    Salary         City  Height_cm  Birth_Year  Product  \
0     Dave  28.0  150000.0      Chicago      150.0        1992   LAPTOP   
1      Eve  28.0   50000.0     new york        NaN        1992   LAPTOP   
2    Carol  35.0   70000.0      Chicago      190.0        1993   LAPTOP   
3      Eve  28.0   50000.0  los angeles      180.0        1990   LAPTOP   
4      Eve   NaN       NaN  los angeles      150.0        1992   Laptop   
..     ...   ...       ...          ...        ...         ...      ...   
145  Alice   NaN  150000.0  Los Angeles      150.0        1994   LAPTOP   
146   Dave   NaN  900000.0     new york      170.0        1993    phone   
147    Eve   NaN  150000.0  Los Angeles      180.0        1990   Laptop   
148   Dave   NaN   70000.0  los angeles      160.0        1993  lap top   
149  Frank  35.0   80000.0     New York        NaN        1994   LAPTOP   

    Purchase_Date  
0             NaN  
1      2022-02-05  
2      2022-03-12  
3      2022-01-01  
4             NaN  
..            ...  
145    2022-03-12  
146           NaN  
147    2022-01-01  
148    2022-02-05  
149    2022-01-01  

[150 rows x 8 columns]

🔹 df.value_counts()

Purpose: Count frequency of each unique value in a column.

In [84]:
# Example of using the value_counts()
df.value_counts()

Name   Age   Salary    City         Height_cm  Birth_Year  Product  Purchase_Date
Carol  35.0  70000.0   Chicago      190.0      1993        LAPTOP   2022-03-12       1
Eve    28.0  50000.0   los angeles  180.0      1990        LAPTOP   2022-01-01       1
Bob    28.0  60000.0   los angeles  190.0      1992        LAPTOP   2022-01-01       1
Carol  35.0  150000.0  Chicago      150.0      1990        phone    2022-02-05       1
             60000.0   los angeles  170.0      1990        Tablet   2022-02-05       1
Dave   30.0  70000.0   los angeles  150.0      1992        LAPTOP   2022-03-12       1
Bob    25.0  50000.0   Chicago      150.0      1992        lap top  2022-01-01       1
Frank  35.0  80000.0   New York     170.0      1993        Phone    2022-03-12       1
Eve    30.0  70000.0   Chicago      180.0      1993        phone    2022-03-12       1
Alice  25.0  150000.0  Chicago      180.0      1985        phone    2022-01-01       1
Carol  25.0  150000.0  new york     180.0      1

# 12. Validation & Quality Checks

🔹 df.isna().sum()

Purpose: Count missing values per column.

In [85]:
# Example of using the isnull().sum()
df.isnull().sum()

Name              0
Age              56
Salary           18
City              0
Height_cm        22
Birth_Year        0
Product           0
Purchase_Date    35
dtype: int64

🔹 df.duplicated().sum()

Purpose: Count total duplicate rows.

In [86]:
# Example of using the duplicated()
df.duplicated().sum()

np.int64(0)

🔹 df.equals()

Purpose: Check if two DataFrames are exactly identical.

In [ ]:
# Example of using the equals()
df2 = df.copy()
print(df.equals(df2))

True


🔹 df.compare()

Purpose: Show differences between two DataFrames.

In [96]:
# Example of using the compare()
df3 = df.copy()
df3['Salary'] = df3['Salary'].fillna(60000)
display(df.compare(df3))

Salary         
      self    other
4      NaN  60000.0
11     NaN  60000.0
18     NaN  60000.0
34     NaN  60000.0
40     NaN  60000.0
46     NaN  60000.0
48     NaN  60000.0
57     NaN  60000.0
59     NaN  60000.0
64     NaN  60000.0
65     NaN  60000.0
68     NaN  60000.0
75     NaN  60000.0
76     NaN  60000.0
104    NaN  60000.0
111    NaN  60000.0
117    NaN  60000.0
137    NaN  60000.0